commit bdf900435a2a5895f8a4f24ea8c9517952fe8fae (refs/changes/40/5040/1)
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2018-11-08T09:06:01-08:00 (11 months ago)
    
    OS-7361 SmartOS initial set up password prompt mangles some patterns

diff --git a/overlay/generic/smartdc/lib/smartos_prompt_config.sh b/overlay/generic/smartdc/lib/smartos_prompt_config.sh
old mode 100644
new mode 100755
index 8674ee8c..5bcf3d25
--- a/overlay/generic/smartdc/lib/smartos_prompt_config.sh
+++ b/overlay/generic/smartdc/lib/smartos_prompt_config.sh
@@ -530,7 +530,7 @@ promptpw()
 	key="$4"
 
 	if [[ -n ${key} ]]; then
-		preset_val=$(getanswer "${key}")
+		preset_val="$(getanswer "${key}")"
 	fi
 
 	trap "" SIGINT
@@ -538,16 +538,14 @@ promptpw()
 		val=""
 		while [ -z "$val" ]; do
 			if [[ -n ${preset_val} ]]; then
-				val=${preset_val}
+				val="${preset_val}"
 			else
 				if [ -z "$def" ]; then
 					printf "%s: " "$1"
 				else
 					printf "%s [enter to keep existing]: " "$1"
 				fi
-				stty -echo
-				read val
-				stty echo
+				IFS='' read -r -s val
 				echo
 			fi
 			if [ -n "$val" ]; then
@@ -561,7 +559,7 @@ promptpw()
 				fi
 			else
 				if [ -n "$def" ]; then
-					val=$def
+					val="$def"
 					return
 				else
 					echo "A value must be provided."
@@ -572,12 +570,10 @@ promptpw()
 		cval=""
 		while [ -z "$cval" ]; do
 			if [[ -n ${preset_val} ]]; then
-				cval=${preset_val}
+				cval="${preset_val}"
 			else
 				printf "%s: " "Confirm password"
-				stty -echo
-				read cval
-				stty echo
+				IFS='' read -r -s cval
 				echo
 			fi
 			[ -n "$cval" ] && break
